atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/CachingTransaction.java

```diff
@@ -137,6 +140,37 @@ public Transaction delegate() {
         return cacheHit;
     }
 
+    @Override
+    public ListenableFuture<Map<Cell, byte[]>> getAsync(TableReference tableRef, Set<Cell> cells) {
+        if (cells.isEmpty()) {
+            return Futures.immediateFuture(ImmutableMap.of());
+        }
+
+        Set<Cell> toLoad = Sets.newHashSet();
+        ImmutableMap.Builder<Cell, byte[]> cacheHitMapBuilder = ImmutableMap.builder();
+        for (Cell cell : cells) {
+            byte[] val = getCachedCellIfPresent(tableRef, cell);
+            if (val != null) {
+                if (val.length > 0) {
+                    cacheHitMapBuilder.put(cell, val);
+                }
+            } else {
+                toLoad.add(cell);
+            }
+        }
+
+        ImmutableMap<Cell, byte[]> cacheHit = cacheHitMapBuilder.build();
```

Comment:
Normal HashMap should correctly publish here, we shouldn't need this additional copy.